### ISSUE TITLE
Timeouts on Poolboy.checkout

### ISSUE THREAD

**Issue by Coffei** [elixir-ecto/db_connection#127]
Hi,
I have recently started seeing the below error in my app
```
14:39:19.261 [error] GenServer AssetMap.Storage.Stats terminating
** (stop) exited in: :gen_server.call(#PID<0.6836.0>, {:checkout, #Reference<0.3038416112.2899312641.205728>, true, 900000}, 5000)
    ** (EXIT) time out
    (db_connection) lib/db_connection/poolboy.ex:112: DBConnection.Poolboy.checkout/3
    (db_connection) lib/db_connection.ex:928: DBConnection.checkout/2
    (db_connection) lib/db_connection.ex:750: DBConnection.run/3
    (db_connection) lib/db_connection.ex:592: DBConnection.prepare_execute/4
    (ecto) lib/ecto/adapters/postgres/connection.ex:86: Ecto.Adapters.Postgres.Connection.execute/4
    (ecto) lib/ecto/adapters/sql.ex:256: Ecto.Adapters.SQL.sql_call/6
    (stdlib) timer.erl:166: :timer.tc/1
    (utils) lib/stats.ex:32: Utils.Stats.timefun/2
Last message: :report_stats
```

It started appearing after I upgraded from Fedora 27 to 28, before the upgrade I never saw such error. The code and dependencies are the same. Still using roughly the same Erlang and Elixir versions,  now using Erlang 20.3 and Elixir 1.6.5, I also tried Elixir 1.6.0 and a couple of versions in between - all behave the same. My guess is that it's something in the system that changed versions.

The error occurs quite randomly, some DB queries go through while some time out. It seems to affect my stats GenServer the most, the stats GenServer is running in background and executes `COUNT(*)` on some of my tables every 5s. It seems to affect 1-2 in every 10 requests. Note the queries are not intensive in any way, when executed directly they return in under 100ms.

The database is Postgres running in a docker. Although I created new containers the image itself hasn't changed, so I assume the database side of things remained intact.

My colleagues are not experiencing the issue, although none of them run Fedora 28. The app is unfortunately not open source.

I am not quite sure where to start debugging this. Could you provide some pointers where to start? Debug logs provide no further information.

**Comment by josevalim** [comment]
Do you have queries or long transactions running that may be blocking the connection pool? Because that is likely blocking the pool is not the query that is causing this particular error, but rather the other queries in your application.

Things you could do to improve:

1. Increase the pool_size
2. Change to DBConnection.Sojourn pool
3. Maybe increase the timeout?

**Comment by Coffei** [comment]
No long transactions, in fact the `COUNT(*)` queries are the only queries that are executed as the issue occurs even when the system is in idle and does nothing else.
Is it somehow possible to see what is holding the connections?

I'll try increasing the pool size and/or the timeout.

**Comment by josevalim** [comment]
If nothing else is happening in the app, then it is not going to be the pool size or timeout. Did you try using another OTP version than 20.3?

**Comment by Coffei** [comment]
OK, it seems to be working fine with OTP 20.0. I'll be testing some more until tomorrow just to be sure. Let me know if I should close this issue if the timeout doesn't appear on 20.0 or if you'd want me to keep this open or provide more info.

**Comment by michalmuskala** [comment]
I've seen some similar reports about OTP 20.3 as well. I'm not really sure what could be causing this.

**Comment by Coffei** [comment]
Hmm.. I can still see the error with 20.0 occasionally, it's definitely much more rare, but it's there.

**Comment by michalmuskala** [comment]
Do you have any metrics measuring the queue and query time from ecto logger? This could help verifying if it's something weird or just connections getting tied up somewhere.

**Comment by fishcakez** [comment]
The timeout is occurring against the connection process not against poolboy, so nothing is blocking the pool. This means queue and query times likely look good (except the timeout failures). This means the pool has a free connection process but that process is not responding in 5 seconds. There is a 900 second timeout, which means the pool will let the calling process use the connection for 900 seconds! That seems a little high. Perhaps increasing the `pool_timeout`, which is currently `5000` by default might help in this situation. The `pool_timeout` is how long to wait for a connection.

However if the connection process can not return a socket within 5 seconds then it suggests something else is not right. My assumption would be that either there is a bug in the connection logic (race condition?) or the database/network is down.

**Comment by fishcakez** [comment]
> It seems to affect 1-2 in every 10 requests

If this is the case then its possible that a single connection process is in a bad state. The default pool size is `10` and we use a FIFO queue, so if there is no or little concurrent pool access a single connection process would get used every 10 requests. If you hit this bug against you could call `:sys.get_state(#PID<0.6836.0>)` on the connection so we can see its state. Note that the state WILL contain sensitive information (password) so make sure to filter that. If in doubt email the output to someone in the thread.

**Comment by Coffei** [comment]
@michalmuskala  The only queries are simple `count(*)` queries, nothing else should be running. Below is the a part of the debug log showing the times.
```
08:48:55.204 [debug] QUERY OK db=64.0ms
SELECT count(*) FROM asset []
 
08:48:55.272 [debug] QUERY OK db=68.0ms queue=0.1ms
SELECT count(*) FROM edge []
 
08:49:25.376 [debug] QUERY OK db=102.8ms queue=0.2ms
SELECT count(*) FROM asset []
 
08:49:25.444 [debug] QUERY OK db=67.3ms queue=0.1ms
SELECT count(*) FROM edge []
```

@fishcakez It really looks like a broken process, I see it's always the same PID in the error message. I tried getting the state but I get timeout
```
** (exit) exited in: :sys.get_state(#PID<0.859.0>)
    ** (EXIT) time out
    (stdlib) sys.erl:305: :sys.send_system_msg/2
    (stdlib) sys.erl:114: :sys.get_state/1
```
This is the process info
```elixir
[
  current_function: {:prim_inet, :recv0, 3},
  initial_call: {:proc_lib, :init_p, 5},
  status: :waiting,
  message_queue_len: 12,
  messages: [
    {:inet_async, #Port<0.14311>, 475, {:ok, <<90, 0, 0, 0, 5, 73>>}},
    {:"$gen_call", {#PID<0.878.0>, #Reference<0.3722681502.2001731585.239489>},
     {:checkout, #Reference<0.3722681502.2001731585.239488>, true, 900000}},
    {:"$gen_cast", {:cancel, #Reference<0.3722681502.2001731585.239488>}},
    {:"$gen_call", {#PID<0.1193.0>, #Reference<0.3722681502.2001731585.240793>},
     {:checkout, #Reference<0.3722681502.2001731585.240792>, true, 900000}},
    {:"$gen_cast", {:cancel, #Reference<0.3722681502.2001731585.240792>}},
    {:"$gen_call", {#PID<0.1199.0>, #Reference<0.3722681502.2001731585.242087>},
     {:checkout, #Reference<0.3722681502.2001731585.242086>, true, 900000}},
    {:"$gen_cast", {:cancel, #Reference<0.3722681502.2001731585.242086>}},
    {:system, {#PID<0.1184.0>, #Reference<0.3722681502.2001731585.243359>},
     :get_state},
    {:"$gen_call", {#PID<0.1203.0>, #Reference<0.3722681502.2001731585.243465>},
     {:checkout, #Reference<0.3722681502.2001731585.243464>, true, 900000}},
    {:"$gen_cast", {:cancel, #Reference<0.3722681502.2001731585.243464>}},
    {:system, {#PID<0.1184.0>, #Reference<0.3722681502.2001731585.243550>},
     :get_state},
    {:system, {#PID<0.1184.0>, #Reference<0.3722681502.2001731585.243662>},
     :get_state}
  ],
  links: [#PID<0.854.0>, #PID<0.855.0>, #PID<0.869.0>, #Port<0.14311>],
  dictionary: [
    "$ancestors": [#PID<0.855.0>, AssetMap.Ecto.Repo.Pool, AssetMap.Ecto.Repo,
     AssetMap.AppSupervisor, #PID<0.851.0>],
    "$initial_call": {DBConnection.Connection, :init, 1}
  ],
  trap_exit: false,
  error_handler: :error_handler,
  priority: :normal,
  group_leader: #PID<0.850.0>,
  total_heap_size: 53194,
  heap_size: 6772, 
  stack_size: 39,
  reductions: 55781,
  garbage_collection: [
    max_heap_size: %{error_logger: true, kill: true, size: 0},
    min_bin_vheap_size: 46422,
    min_heap_size: 233,
    fullsweep_after: 65535,
    minor_gcs: 5
  ],
  suspending: []
]
```

**Comment by michalmuskala** [comment]
Interesting. It looks like the process is blocking in a socket receive for data that never comes... Did you try calling it after some time to see if this changes or is the process stuck doing just this? Could you also call `Process.info(pid, :current_stacktrace)` on the process when this happens? This should show us a bit more about what exactly the process is doing and maybe also which receive call exactly that is.

**Comment by Coffei** [comment]
I have the app running for about 30 minutes and the process is still stuck. In fact, another process that didn't seem to be stuck at first got stuck after a while, it's spreading like a disease :D.

The stack trace is the same for both processes
```elixir
{:current_stacktrace,
 [
   {:prim_inet, :recv0, 3, []},
   {Postgrex.Protocol, :msg_recv, 4,
    [file: 'lib/postgrex/protocol.ex', line: 1956]},
   {Postgrex.Protocol, :ping_recv, 4,
    [file: 'lib/postgrex/protocol.ex', line: 1705]},
   {DBConnection.Connection, :handle_info, 2,
    [file: 'lib/db_connection/connection.ex', line: 373]},
   {Connection, :handle_async, 3, [file: 'lib/connection.ex', line: 810]},
   {:gen_server, :try_dispatch, 4, [file: 'gen_server.erl', line: 616]},
   {:gen_server, :handle_msg, 6, [file: 'gen_server.erl', line: 686]},
   {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 247]}
 ]}
```

I also checked the connections from postgres and all seem to be idle.

**Comment by michalmuskala** [comment]
From that it looks like the database is not replying to ping requests. Did you also change version of postgres when you updated your system?

**Comment by Coffei** [comment]
Nope, the database definitely didn't change versions as its running in docker and the image is the same. It must be something in the system that changed versions..

**Comment by josevalim** [comment]
Thanks @Coffei for all the help distilling this. Since you can isolate the issue to that command, do you think you can reproduce it in a smaller application? Maybe even in one of the examples in our examples folder?

**Comment by Coffei** [comment]
I've put together a small reproducer using only Postgrex and db_connection with poolboy - https://github.com/Coffei/timeout_reproducer. 

The error usually appears after few tries, sometimes I have to re-run the app a few times. The error is very similar
```
09:01:32.446 [error] Task #PID<0.283.0> started from #PID<0.276.0> terminating
** (stop) exited in: :gen_server.call(#PID<0.179.0>, {:checkout, #Reference<0.1346686728.617611267.217656>, true, 5000}, 5000)
    ** (EXIT) time out
    (db_connection) lib/db_connection/poolboy.ex:112: DBConnection.Poolboy.checkout/3
    (db_connection) lib/db_connection.ex:928: DBConnection.checkout/2
    (db_connection) lib/db_connection.ex:750: DBConnection.run/3
    (db_connection) lib/db_connection.ex:1141: DBConnection.run_meter/3
    (db_connection) lib/db_connection.ex:592: DBConnection.prepare_execute/4
    (postgrex) lib/postgrex.ex:146: Postgrex.query/4
    (timeout_reproducer) lib/app.ex:28: anonymous fn/1 in TimeoutReproducer.App.test/0
    (elixir) lib/task/supervised.ex:88: Task.Supervised.do_apply/2
Function: #Function<1.76782678/0 in TimeoutReproducer.App.test/0>
    Args: []
```

I also see the following error, which appears independently of the one above.
```
09:01:32.437 [error] Postgrex.Protocol (#PID<0.177.0>) disconnected: ** (DBConnection.ConnectionError) client #PID<0.288.0> timed out because it checked out the connection for longer than 5000ms
```

**Comment by Coffei** [comment]
Also recorded asciinema - https://asciinema.org/a/jWGGvIE12qve7mCTHKqTBgw7I it appears on the first try so I restart the app and it doesn't come up any more.

**Comment by Coffei** [comment]
I found that increasing the pool size actually made things worse as there was a bigger chance a process enters the bad state. Increased the pool size in the reproducer to 100, so it should be easier to reproduce.

I've been testing on clean VMs and I could reproduce this only in F28, both F27 and Ubuntu 18.04 were working well, so I assume it's something in the system that's interfering. Note it was much more rare in the VMs, in a slower VirtualBox it took about 30 minutes to reproduce, so I'd say it's some race condition that only appears on sufficiently fast setups.

**Comment by fishcakez** [comment]
@Coffei my last message was suggesting it is a bug in Erlang/OTP can you share the exact OTP versions of all thoses?

**Comment by Coffei** [comment]
On my local box I tried with Erlang 20.0 and 20.3 (from source and from repository), on the VMs its 20.3 for F28, 19.3 for F27 and 20.2 for the Ubuntu VM. Kinda inconsistent, but we tried in pair with @lubosvesely, me running F28 and he Ubuntu (18 or 16) with the exact same versions of Erlang and Elixir (down to patch versions), same database and the same reproducer. And @lubosvesely still wasn't able to reproduce the error. And before I upgraded the PC last week and was running F27 with Erlang 20.x and Elixir 1.6.x I also didn't see the issue.

As far as I see it, the system must play some role in it.

**Comment by fishcakez** [comment]
@Coffei just to be clear the output you showed in https://github.com/elixir-ecto/db_connection/issues/127#issuecomment-396833857 should not be possible without a bug in Erlang/OTP. Of course it is possible that the race condition that causes it is more likely to occur on your machine. The reason for the exact OTP versions is so we can report this upstream if we get another report.

**Comment by nixpulvis** [comment]
I am also effected by this I believe. At random times (generally not *right* after starting the server) I get the following error, which is generally worked through by making a few more requests, one of them eventually works.

```
** (exit) exited in: :gen_server.call(#PID<0.789.0>, {:checkout, #Reference<0.1448561368.3619946498.18872>, true, 15000}, 5000)
    ** (EXIT) time out
```

Looking at the `Process.info` we get the following.

```
[
  current_function: {:prim_inet, :recv0, 3},
  initial_call: {:proc_lib, :init_p, 5},
  status: :waiting,
  message_queue_len: 3,
  messages: [
    {:inet_async, #Port<0.46938>, 1087, {:ok, <<90, 0, 0, 0, 5, 73>>}},
    {:"$gen_call", {#PID<0.1057.0>, #Reference<0.1448561368.3619946498.18873>},
     {:checkout, #Reference<0.1448561368.3619946498.18872>, true, 15000}},
    {:"$gen_cast", {:cancel, #Reference<0.1448561368.3619946498.18872>}}
  ],
  links: [#PID<0.786.0>, #PID<0.787.0>, #PID<0.804.0>, #Port<0.46938>],
  dictionary: [
    "$ancestors": [#PID<0.787.0>, Auth.Repo.Pool, Auth.Repo, Auth.Supervisor,
     #PID<0.783.0>],
    "$initial_call": {DBConnection.Connection, :init, 1}
  ],
  trap_exit: false,
  error_handler: :error_handler,
  priority: :normal,
  group_leader: #PID<0.782.0>,
  total_heap_size: 3573,
  heap_size: 2586,
  stack_size: 39,
  reductions: 14872,
  garbage_collection: [
    max_heap_size: %{error_logger: true, kill: true, size: 0},
    min_bin_vheap_size: 46422,
    min_heap_size: 233,
    fullsweep_after: 65535,
    minor_gcs: 12
  ],
  suspending: []
]
```

System information

```
Erlang/OTP 20 [erts-9.2.1] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:10] [hipe] [kernel-poll:false]
```

**Comment by fishcakez** [comment]
@Coffei do you have `hipe` compiled code enabled? It would be in the output above with `[hipe]` if so.

**Comment by isaiahdw** [comment]
We've noticed the same thing recently. No changes to the database (postgresql 9.6.4 on Heroku), but we did upgrade Elixir and Erlang/OTP recently. I'll do some more debugging, but it seems related to the Elixir/Erlang changes. 

We also noticed that increasing the pool size made this problem happen more often.

**Comment by fishcakez** [comment]
@isaiahdw could you include the before/after Erlang/OTP versions?

**Comment by isaiahdw** [comment]
@fishcakez, sorry about the slow reply. We upgraded from OTP 20.1.6 to 20.3.8. Although I'm less sure it is related to the OTP version now. We've been running the last two days without any timeouts and we're still on OTP 20.3.

I'll post more details here if I'm able to reproduce or start seeing errors again.

**Comment by fishcakez** [comment]
@isaiahdw did you change OS version?

**Comment by isaiahdw** [comment]
@fishcakez, no OS changes. This is an application hosted on Heroku, so I can't be 100% sure about changes that I might not be aware of, but the only change we made was the Elixir and OTP updates.

**Comment by Kabie** [comment]
Today I upgraded the erlang to 20.3.8-1 on Archlinux, and hit this very issue.

**Comment by oestrich** [comment]
I just got this issue with Erlang 21.0 and Elixir 1.6.6. I was on Erlang 20.2 and Elixir 1.6.5 previously. Arch and `hipe` enabled for both version sets.

**Comment by josevalim** [comment]
@nixpulvis what is your OS?

**Comment by nixpulvis** [comment]
@josevalim 

```
$ uname -a
Linux milik 4.16.9-1-ARCH #1 SMP PREEMPT Thu May 17 02:10:09 UTC 2018 x86_64 GNU/Linux
```

**Comment by semaphase** [comment]
Having the same issue

``$ uname -a``
``Linux user 4.17.2-1-ARCH #1 SMP PREEMPT Sat Jun 16 11:08:59 UTC 2018 x86_64 GNU/Linux``

**Comment by josevalim** [comment]
Thanks everybody for the info so far.

I have tried to reproduce this issue on a Virtual Box VM but failed. But probably due to my lack of knowledge of both Virtual Box and Fedora.

Can we make this easier to reproduce? Maybe someone can use Docker and get @Coffei's [sample application](https://github.com/Coffei/timeout_reproducer) inside Fedora 28 (or any other OS that reproduces the issue) and make it easier to reproduce via some command line instructions?

Can somebody please look into this? Thanks!

**Comment by Coffei** [comment]
I've added a dockerfile to the reproducer app, based on Fedora 28. You can run it as follows

1. Run database
`docker run --name db -e POSTGRES_USER=assetmap -e POSTGRES_PASSWORD=assetmap -e POSTGRES_DATABASE=assetmap -d --net=host postgres:10`
2. Run the reproducer 
`docker run --net=host -it coffei/timeout_reproducer`
This will open up iex console of the app.
3. Now you can execute `TimeoutReproducer.App.test` multiple times as I mentioned previously. I tried it myself and reproducer on a second try.

Right now the image is still being built, should be done soon - https://hub.docker.com/r/coffei/timeout_reproducer/builds/

**Comment by karlseguin** [comment]
I don't have much to add, except to say we've had this same problem for over a year. It's been very infrequent (once every month or so), so we're only now starting to look seriously into it. We're running in EC2, connecting to AWS, Docker on Ubuntu.

I have two immediate questions. 

1 - Is the ping being called on each connection checkout? Have people found this to be worthwhile? I believe that neither Go's lib/pg or jackc/pgx, nor Node's PG library bother to ping on checkout. Frankly I find it of dubious value since the connection could simply die after the ping but before the real query is sent.

2 - Postgrex is sending a SYNC command. Where does that idea come from? Googling, it seems that drivers that do support ping, do a `select 1`  (or `select  'postgrex ping`) using a simple query. For example, this reply from the psycopg2 developer says to just issue a `select 1` if you must  https://github.com/psycopg/psycopg2/issues/419 (psycopg2 also doesn't support pings)

**Comment by josevalim** [comment]
Thank you @Coffei! I'm downloading it now.

@karlseguin we only ping on `idle_timeout`, which defaults to 1 second. If this is the root cause, we can try a couple things:

1. Increase the idle_timeout to 10_000 (10 seconds) and see if the reduces the frequency.
2. Use another pool besides DBConnection.Poolboy. For example, try DBConnection.Sojourn.

**Comment by josevalim** [comment]
@Coffei Unfortunately I cannot reproduce the issue even with docker. I have it running the query periodically for the last 10 minutes and nothing yet. :( Any ideas?

Meanwhile, can you please try the two suggestions above (idle_timeout and changing the pool) and let us know if it changes anything?

**Comment by Coffei** [comment]
No idea, as I mentioned earlier it seems quite random and people with very similar setups were not able to reproduce it while I get it all the time. What worked for me was to restart the app if it didn't show the error after 1-2 tries, as the connections seemed to be often corrupted right after they were created.

I'll try the two suggestions and report back.

**Comment by Coffei** [comment]
I can see the issue with both suggestions. The idle timeout doesn't seem to have any effect, the `Sojourn` pool seems to swallow the errors a bit, but the error below still shows up while some queries don't finish properly, so I'd say it's also the same.
```
13:54:43.467 [error] Postgrex.Protocol (#PID<0.283.0>) disconnected: ** (DBConnection.ConnectionError) client #PID<0.1013.0> timed out because it checked out the connection for longer than 5000ms
```

**Comment by arjan** [comment]
I also have this issue in my application. Recently upgraded to 1.6.6 and OTP 21 (arch). Postges 10.4. 

Weirdly enough it only occurs on one repo, pretty consistently (my app has 2). I'll try to reproduce it with the sample app.

**Comment by arjan** [comment]
Yes, I have the same issue in this test app, happens after a few times already. More and more workers are being blocked in the `ping_recv` that is supposed to return quickly.

**Comment by josevalim** [comment]
@arjan do you reproduce it locally? If not, which OS are you on?

**Comment by arjan** [comment]
Yes locally, laptop w/ Arch linux

**Comment by arjan** [comment]
Checking if downgrading to OTP 20.3 fixes it...

**Comment by arjan** [comment]
This is the process info of the worker that hangs:

```
[       
  current_function: {:prim_inet, :recv0, 3},
  initial_call: {:proc_lib, :init_p, 5},
  status: :waiting,
  message_queue_len: 2,
  links: [#PID<0.198.0>, #Port<0.15>, #PID<0.214.0>, #PID<0.197.0>],
  dictionary: [
    "$initial_call": {DBConnection.Connection, :init, 1},
    "$ancestors": [#PID<0.198.0>, TestPool, TimeoutReproducer.Supervisor,
     #PID<0.195.0>]
  ],
  trap_exit: false,
  error_handler: :error_handler,
  priority: :normal,
  group_leader: #PID<0.194.0>,
  total_heap_size: 2208,
  heap_size: 1598,
  stack_size: 39,
  reductions: 4672,
  garbage_collection: [
    max_heap_size: %{error_logger: true, kill: true, size: 0},
    min_bin_vheap_size: 46422,
    min_heap_size: 233,
    fullsweep_after: 65535,
    minor_gcs: 12
  ],
  suspending: []
]
```

```
iex(7)> Process.info(pid(0,201,0), :current_stacktrace)
{:current_stacktrace,
 [
   {:prim_inet, :recv0, 3, []},
   {Postgrex.Protocol, :msg_recv, 4,
    [file: 'lib/postgrex/protocol.ex', line: 1985]},
   {Postgrex.Protocol, :ping_recv, 4,
    [file: 'lib/postgrex/protocol.ex', line: 1734]},
   {DBConnection.Connection, :handle_info, 2,
    [file: 'lib/db_connection/connection.ex', line: 373]},
   {Connection, :handle_async, 3, [file: 'lib/connection.ex', line: 810]},
   {:gen_server, :try_dispatch, 4, [file: 'gen_server.erl', line: 637]},
   {:gen_server, :handle_msg, 6, [file: 'gen_server.erl', line: 711]},
   {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 249]}
 ]}
```

Have to go now, will dig further tonight.

**Comment by luisgabrielroldan** [comment]
I'm having the same  issue. Downgrades are not working for me.
I tried downgrading OTP to 20 and older elixir versions but still happens.

I used asdf to setup the older versions that I know it was working fine on my project:
```
Erlang/OTP 20 [erts-9.2] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:10] [hipe] [kernel-poll:false]

Elixir 1.5.2
```
... and still happens.

I'm using Arch Linux too...
This is happening after a system upgrade (I did it on Saturday but I have been putting it off for weeks).

**Comment by pera** [comment]
@josevalim I just created a simple Phoenix application that reproduce this issue on my Arch Linux laptop:

**Comment by arjan** [comment]
Weird, I cant reproduce it on OTP 20. Don't think the Elixir version has much to do with it, BTW.

**Comment by luisgabrielroldan** [comment]
@arjan It's happening to me on OTP 20 and 21. As you said this doesn't looks like an Elixir bug, I think is related to a new library release...

The only way I found to continue working today was edit my mirrorlist file and use an snapshot from 15-04-2018 (maybe a newer one could work too):

```
Server=https://archive.archlinux.org/repos/2018/04/15/$repo/os/$arch
```

**Comment by pera** [comment]
@josevalim I believe I found the issue with your testing environment: your VM was probably using one single core (this is the default in VirtualBox, for instance). I created an Arch Linux instance and wasn't able to reproduce the issue until I changed the number of CPUs from 1 to 4.

Let me know if you need this appliance.

**Comment by arjan** [comment]
@pera would be nice if you could share it!

**Comment by arjan** [comment]
Issue update: we narrowed the stuck db workers down to the `prim_inet:recv0` function in OTP where it receives on `{inet_async, S, Ref _}` but the `Ref` variable is off-by-one with a `inet_async` message in the processes inbox.

**Comment by josevalim** [comment]
@arjan Btw, did you check if the port that we get stuck on is still alive? Something like this would do it:

```iex
iex(3)> Port.monitor :erlang.list_to_port('#Port<0.64>')
#Reference<0.1283475765.2787377154.58319>
iex(4)> flush
{:DOWN, #Reference<0.1283475765.2787377154.58319>, :port, #Port<0.64>, :noproc}
:ok
```

**Comment by michalmuskala** [comment]
@arjan that's a great find. This confirms it's an issue in OTP itself and we should report it to https://bugs.erlang.org/ (unfortunately it's down right now 😩).

**Comment by arjan** [comment]
@josevalim will do that to confirm. @michalmuskala yes I'm first trying to find a way to reproduce it reliably

**Comment by arjan** [comment]
OK it seems to have to do with a compiler gcc optimization. If I compile `inet_drv.c` without optimization (`-O0`), it does not happen, with `-O3` (the default) it does happen. (with -O2 it happens as well, with -O1 it does not, so there is the cutoff point).

GCC version: 8.1.1 20180531

It would be nice if somebody could reproduce my findings :-)

**Comment by arjan** [comment]
@josevalim I got it reproduced on a digitalocean machine (Ubuntu latest, GCC 8.1), I mailed you the details. If somebody else wants the access to the machine to investigate, ping me.

**Comment by nixpulvis** [comment]
I've downloaded both OTP and Elixir from GitHub on master, and compiled them each from source. First OTP, following the instructions in the README. I've confirmed that it worked with `erl`. Then after prepending my `$PATH` with it's bin directory, I compiled Elixir from source and confirmed it worked with `iex`,  ensuring the versions were updated from my system's installation. Then as I did with Erlang/OTP, I prepended it's bin to my `$PATH`.

With this I tested https://github.com/Coffei/timeout_reproducer, and verified the issue still occurs. Then I attempted to set the `-O0` flag by changing `OPT_LEVEL` inside `erts/emulator/Makefile.in` is this correct? If not what Makefile value needs to be changed. With this updated compilation I can still reproduce the issue.

Thoughts @arjan?

**Comment by arjan** [comment]
@nixpulvis thanks for the details. Elixir doesnt need to recompiled indeed.
What I did was `CFLAGS=-O0 ./configure ...` and the `CFLAGS=-O0 make -j4`. Followd by a new `make install` of course.

Also checking the log output and just recompiling `inet_drv.c` followed by a `make install` again did it for me.

Full gcc commandline for reference:
```
gcc -Werror=undef -Werror=implicit -Werror=return-type  -g -O3 -I/home/arjan/src/otp/erts/x86_64-unknown-linux-gnu   -fno-tree-copyrename  -D_GNU_SOURCE  -DHAVE_CONFIG_H -Wall -Wstrict-prototypes -Wmissing-prototypes -Wdeclaration-after-statement -DUSE_THREADS -D_THREAD_SAFE -D_REENTRANT -DPOSIX_THREADS -D_POSIX_THREAD_SAFE_FUNCTIONS   -DLIBSCTP= -Ix86_64-unknown-linux-gnu/opt/smp -Ibeam -Isys/unix -Isys/common -Ix86_64-unknown-linux-gnu -Ipcre -Ihipe -I../include -I../include/x86_64-unknown-linux-gnu -I../include/internal -I../include/internal/x86_64-unknown-linux-gnu -Idrivers/common -Idrivers/unix -c drivers/common/inet_drv.c -o obj/x86_64-unknown-linux-gnu/opt/smp/inet_drv.o
```
(run from the *erts/emulator* directory in the OTP source tree)

**Comment by arjan** [comment]
changing `OPT_LEVEL` seems also OK btw. I will doublecheck this, I might be wrong as well. Mailed you the server details, dont have much time tonight anymore.

**Comment by arjan** [comment]
@Coffei can you check whether your OTP is compiled with GCC 8.1? It might be related to the compiler, not to the OTP version…

You can check it like this:
```
# strings /usr/lib/erlang/erts-10.0/bin/beam.smp|grep GCC
```

**Comment by nixpulvis** [comment]
@arjan I've rerun everything locally following your commands more directly, still getting errors, though they *might* be happening less. It's really hard to tell. I'm personally not 100% sure how Elixir even integrates with Erlang, so I'm flying a bit blind here (just being honest). It could be the case I didn't tell Elixir to use the correct version of Erlang.

**Comment by pera** [comment]
@arjan here is the VirtualBox appliance with Arch Linux:
https://drive.google.com/file/d/1U8zcbdWYZRwvnKoJ7E6TAjEKM2Pq23Ah/view
Login with root and then go to ~/db_connection_issue127, there you will find a README file with instructions to reproduce this issue. After a few tries you should see something like this:
![screenshoot](https://i.imgur.com/UUwVwJS.png)

**Comment by netsudo** [comment]
@arjan 

Experiencing the same issue as well, Erlang/OTP 20 (although I've just recently downgraded from OTP 21 to try to mitigate this issue) with Elixir 1.6.5. Running strings gave me 

`#define ETHR_GCC_HAVE_DW_CMPXCHG_ASM_SUPPORT 1
/* #undef ETHR_GCC_HAVE_SSE2_ASM_SUPPORT */
#define ETHR_HAVE_GCC_ASM_ARM_DMB_INSTRUCTION 0
#define ETHR_HAVE_GCC___ATOMIC_BUILTINS 1
/* #undef ETHR_PREFER_GCC_NATIVE_IMPLS */
#define ETHR_TRUST_GCC_ATOMIC_BUILTINS_MEMORY_BARRIERS 0
GCC: (GNU) 8.1.0
GCC: (GNU) 8.1.1 20180531
`
My project never experiences issue on my desktop though, only on my laptop.

**Comment by Coffei** [comment]
@arjan Yes, even though I personally downgraded to GCC 7.3 (had some problems with 8.1), my OTPs seem to be compiled with 8.1.

**Comment by arjan** [comment]
Thanks for all the info, I'm writing a bug report for the OTP team now.

**Comment by josevalim** [comment]
Thank you @pera! I was able to reproduce it in your VM with @Coffei's app.

**Comment by arjan** [comment]
I created the but report: https://bugs.erlang.org/browse/ERL-654

**Comment by josevalim** [comment]
@arjan I can also reproduce it on MacOS X with GCC 8.1. :D

**Comment by pera** [comment]
I tried [John Högberg's patch](https://github.com/jhogberg/otp/commit/e27b98b353df6b4d02d0180fd6ffa86519b41917) on OTP's master branch and the race condition seems completely fixed :)

**Comment by Coffei** [comment]
Wonderful news, seems like we have a fix coming to OTP soon. Thanks to everybody involved!

**Comment by josevalim** [comment]
Thank you @Coffei for getting this started, providing information and a minimal app!

**Comment by josevalim** [comment]
This is fixed on OTP 21.0.2. Thanks @Coffei for guiding this and everyone who helped with information, debugging, etc. 🎉 I am not familiar with how archlinux works but it would be really appreciated if somebody makes sure that the Erlang/OTP package in their repository gets updated too.

**Comment by tcoopman** [comment]
It's already flagged as out of date on the archlinux, so I guess it won't take long for the new version to drop: https://www.archlinux.org/packages/community/x86_64/erlang/

**Comment by ktec** [comment]
Seriously :+1: :100: :heart: :blue_heart: :green_heart: :yellow_heart: :rabbit: :rabbit: :+1: to all'a you. Thank you :1st_place_medal:

**Comment by dimitarvp** [comment]
Just found this and it helped me. Thanks so much to everybody!

**Comment by kamciokodzi** [comment]
Hello! 
Are there any chances that it stills occurs or came back on otp 22.2.6? 
We have encountered this problem recently and it's causing a lot of problems - even Application downtimes. It is happening quiet randomly I would said but I'm not 100% sure. Everything seems to be working fine and then :boom: suddenly `response times` are going up from 20-50ms to even 20000ms. I have checked for slow queries but didn't found any and I'm quiet concerned. Queue time was high but it was in the same moment when crash happend so I'm not sure if that was a reason. 

We are building release on Docker from `erlang:22.2.6` image with elixir `1.10.1` and then running release on `debian:buster-slim`. Before that we run app on `erlang:22.0` and elixir `1.9.0` for building release and run it on `debian:stretch-slim` but it was also a problem so I'm not sure if that's something with our code or some OTP issue again 😢 

I would love to hear some advices from you guys. 

```
exit: ** (exit) exited in: :gen_server.call(Postgres.Repo.Pool, {:checkout, #Reference<0.1066510293.4072931345.103161>, true}, 5000)
    ** (EXIT) time out
  File "gen_server.erl", line 223, in :gen_server.call/3
  File "/app/deps/poolboy/src/poolboy.erl", line 63, in :poolboy.checkout/3
  File "lib/db_connection/poolboy.ex", line 41, in DBConnection.Poolboy.checkout/2
  File "lib/db_connection.ex", line 928, in DBConnection.checkout/2
  File "lib/db_connection.ex", line 750, in DBConnection.run/3
  File "lib/db_connection.ex", line 592, in DBConnection.prepare_execute/4
  File "lib/ecto/adapters/postgres/connection.ex", line 86, in Ecto.Adapters.Postgres.Connection.execute/4
  File "lib/ecto/adapters/sql.ex", line 256, in Ecto.Adapters.SQL.sql_call/6
```

**Comment by michalmuskala** [comment]
Just as a data point `db_connection` does not use poolboy since version 2.0 released around 1.5 years ago.

**Comment by kamciokodzi** [comment]
@michalmuskala yea i know but we are still using Ecto 2.2 which is using `poolboy 1.5` in that specific version. 
So basically are you saying that the only way to fix that is to bump ecto?
